File: C++/cpu_cores.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstdlib>
#include <unistd.h>

std::string get_os_name() {
    std::ifstream file("/etc/os-release");
    std::string line;
    if (file.is_open()) {
        while (std::getline(file, line)) {
            if (line.rfind("PRETTY_NAME=", 0) == 0) {
                std::string val = line.substr(12);
                if (!val.empty() && val.front() == '"') val.erase(0, 1);
                if (!val.empty() && val.back() == '"') val.pop_back();
                return val;
            }
        }
    }
    return "Linux";
}

std::string get_cpu_model() {
    std::ifstream file("/proc/cpuinfo");
    std::string line;
    if (file.is_open()) {
        while (std::getline(file, line)) {
            if (line.rfind("model name", 0) == 0) {
                size_t colon = line.find(':');
                if (colon != std::string::npos) {
                    std::string model = line.substr(colon + 1);
                    // Trim leading spaces
                    size_t start = model.find_first_not_of(" \t");
                    if (start != std::string::npos) model = model.substr(start);
                    return model;
                }
            }
        }
    }
    return "Generic CPU";
}

int main(int argc, [[maybe_unused]] char *argv[]) {
    if (argc > 1) {
        if (const char* httpd = std::getenv("SERVER_SOFTWARE")) {
            std::cout << "HTTP Server is " << httpd << "\n";
        }
        std::cout << "\n OS:  " << get_os_name() << "\n";
        std::cout << "CPU:  " << get_cpu_model() << "\n\n";
    } else {
        long num_cpus = sysconf(_SC_NPROCESSORS_ONLN);
        std::cout << num_cpus << "\n";
    }
    return 0;
}

File: Rust/src/bin/cpu_cores.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};

fn get_os_name() -> String {
    if let Ok(file) = File::open("/etc/os-release") {
        let reader = io::BufReader::new(file);
        for line in reader.lines().map_while(Result::ok) {
            if line.starts_with("PRETTY_NAME=") {
                let mut val = line["PRETTY_NAME=".len()..].to_string();
                // Strip structural quotes
                if val.starts_with('"') { val.remove(0); }
                if val.ends_with('"') { val.pop(); }
                return val;
            }
        }
    }
    "Linux".to_string()
}

fn get_cpu_model() -> String {
    if let Ok(file) = File::open("/proc/cpuinfo") {
        let reader = io::BufReader::new(file);
        for line in reader.lines().map_while(Result::ok) {
            if line.starts_with("model name") {
                if let Some(colon_idx) = line.find(':') {
                    return line[colon_idx + 1..].trim().to_string();
                }
            }
        }
    }
    "Generic CPU".to_string()
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() > 1 {
        if let Ok(httpd) = env::var("SERVER_SOFTWARE") {
            println!("HTTP Server is {}", httpd);
        }
        println!("\n OS:  {}", get_os_name());
        println!("CPU:  {}\n", get_cpu_model());
    } else {
        // Query cores directly from standard system config thread manager
        let num_cpus = num_cpus::get();
        println!("{}", num_cpus);
    }
}

